### ISSUE TITLE
Learning L0 foundation — canonical player-safe dataset / BC artifact / LearnedPolicy vertical slice

### ISSUE THREAD

**Issue by lisbun** [lisbun/lisjong#184]
## Purpose

Implement the first **lisjong-native Learning vertical slice** under lisbun/lisjong-project#75.

Use the offense-foundation track (lisbun/lisjong-project#73) as the concrete consumer, while treating the existing lisjong-arena #331/#332 implementation as reference/evidence rather than bulk-migrating it.

The target is a minimal end-to-end path:

```text
versioned player-safe source record
    ↓
lisjong-owned feature / dataset materialization
    ↓
bounded BC training
    ↓
immutable model artifact
    ↓
lisjong-owned LearnedPolicy inference
    ↓
Arena consumes the candidate through a stable public Policy interface
```

Refs:
- lisbun/lisjong-project#75
- lisbun/lisjong-project#77
- lisbun/lisjong-project#73
- lisbun/lisjong-project#74
- lisbun/lisjong-arena#331
- lisbun/lisjong-arena#332
- lisbun/lisjong-arena#342
- lisbun/lisjong-arena#346

## Architecture boundary

`lisjong` is the canonical owner of:

- player-safe model-facing representation;
- Learning dataset semantics/materialization;
- teacher/label interpretation where AI semantics are involved;
- training objective / trainer;
- model artifact contract;
- LearnedPolicy model loading / inference;
- intrinsic Learning diagnostics.

`lisjong-arena` remains the owner of:

- environment execution / observation;
- Arena-owned seed/population provenance;
- matchup / rotation / holdout;
- game-strength evaluation;
- external/live benchmark execution;
- evaluation-specific artifacts.

Do not introduce a reverse dependency from `lisjong-engine` to AI/Learning code.

### Seed / population provenance boundary

Follow the project-wide seed allocation / provenance contract in lisbun/lisjong-project#77.

For this Issue's first vertical slice:

```text
Arena-produced source population
    -> lisjong-arena owns allocation / freshness / non-overlap provenance
    -> lisjong consumes and preserves the source allocation / population identity

lisjong dataset / training artifact
    -> binds the source population identity
    -> records training RNG seed/config as artifact provenance

lisjong-owned execution population
    -> not introduced by this Issue
```

Therefore #184 does **not** implement a lisjong seed-allocation ledger merely to consume the #342 source record. The Arena-owned source population remains governed by the Arena ledger / provenance path (lisbun/lisjong-arena#346).

If a future lisjong capability becomes the producer of a stochastic game/simulation population — for example self-play, training-data generation, or Learning-specific simulation — create a concrete lisjong implementation child under project #77 and make that producer own its allocation ledger / seed-domain provenance.

Do not silently treat a training RNG seed as equivalent to an Arena hanchan population seed. Cross-owner non-overlap is enforced only when a scientific protocol explicitly requires it, using the #77 provenance contract.

## First-slice scope

Implement only capabilities required to reproduce a bounded offense-foundation learner from a reusable player-safe source record.

### 1. Source-record consumer

Consume an explicit versioned source-record schema rather than an Arena-encoded feature tensor as canonical input.

Initial producer: lisbun/lisjong-arena#342.

Requirements:

- strict schema/version validation;
- fail closed on unknown/inconsistent versions;
- preserve source population / hanchan / decision provenance;
- validate legal action and selected canonical action consistency.

### 2. Canonical player-safe feature representation

Define a lisjong-owned feature identity suitable for Learning.

Requirements:

- derived only from player-safe PolicyInput-equivalent information;
- explicit stable version/fingerprint;
- deterministic materialization;
- no privileged hidden information;
- no reuse/renaming of historical `arena-policy-input-feature-v1` identity.

Do not assume the historical 8204-dimensional Arena representation is the final canonical lisjong representation.

### 3. Dataset materialization

Define a versioned dataset contract that binds:

- source-record identity;
- feature identity;
- action vocabulary identity;
- teacher/label semantics;
- ordered source population;
- split membership where applicable.

Dataset creation must be reproducible and strict-read validated.

### 4. Bounded BC trainer

Provide the minimum training path needed for the offense L0 use case.

Keep model/training scope deliberately narrow. No generic experiment framework or hyperparameter search.

Training semantics and executable entry point belong here in lisjong.

### 5. Model artifact

Use immutable, repository-external model artifacts.

Bind at minimum:

- source/dataset identity;
- feature fingerprint;
- action vocabulary fingerprint;
- model architecture/config;
- optimizer/training config;
- RNG/training seed;
- checkpoint digest;
- relevant lisjong revision/semantic identity.

No arbitrary callable/factory serialization. No silent overwrite. Fail closed on inconsistent load.

### 6. LearnedPolicy inference

Expose a lisjong-owned Policy implementation/factory suitable for Arena execution.

Requirements:

- importable top-level factory; no lambda/closure dependency;
- fresh Policy instance per game/seat where required by existing Policy lifecycle;
- canonical InternalAction output;
- preserve normal legal-action validation;
- silent heuristic fallback is forbidden.

## Runtime dependency boundary

Keep existing lisjong core importable without a heavy ML framework.

Target structure:

```text
core policy / hand / belief / action vocabulary
    -> no ML runtime required

learning / learned inference
    -> optional extra + lazy import
```

CI must include:

1. pinned CPU ML job for Learning code;
2. regression proving core imports/tests succeed without the ML extra.

## Relationship to the existing Arena experiment

The current #331/#332 experiment remains a locked Arena experiment.

This Issue does not rewrite its:

- teacher;
- feature identity;
- thresholds;
- scientific population;
- artifact identity;
- result classification.

Historical Arena implementation may be inspected for invariants, failure modes, artifact discipline, and test cases.

New lisjong-native identities must be distinct.

## Evaluation seam

Arena must be able to consume the produced LearnedPolicy without knowing trainer internals.

The model/Policy contract should be sufficient for later controlled Arena evaluation, but this Issue does not own game-strength conclusions.

## Suggested implementation sequence

```text
L0a
  source-record schema consumer
  + provenance validation

L0b
  canonical feature identity
  + dataset materializer

L0c
  bounded BC trainer
  + immutable artifact

L0d
  LearnedPolicy inference
  + Arena-compatible public seam
```

Keep these as slices of one concrete vertical use case. Split further only if implementation size demands it.

## Acceptance criteria

- [x] reusable source-record input is consumed without depending on Arena feature tensors as canonical input
- [x] source population / allocation provenance from the Arena producer is preserved and bound into downstream dataset/artifact identity without re-owning the Arena allocation
- [x] lisjong-owned player-safe feature identity is versioned/fingerprinted
- [x] dataset materialization is deterministic and strict-read validated
- [x] action vocabulary uses the canonical lisjong vocabulary contract
- [x] bounded BC training can produce an immutable external artifact
- [x] artifact load fails closed on identity/config mismatch
- [x] LearnedPolicy can perform inference through the normal lisjong Policy contract
- [x] silent fallback is absent
- [x] core lisjong remains ML-runtime independent
- [x] ML and no-ML CI paths are both tested
- [x] Arena can execute the Policy through a stable public interface
- [x] historical Arena feature/dataset/checkpoint identities remain unchanged

## Non-goals

- bulk migration of lisjong-arena ML code;
- deleting historical Arena research modules;
- generic ML platform construction;
- self-play / RL;
- defense / wait-shape auxiliary supervision;
- calls / open-hand strategy;
- Champion promotion;
- changing heuristic teacher behavior;
- changing #331/#332 locked scientific protocol.

## Completion outcome

lisjong can build and serve one bounded LearnedPolicy end to end from a versioned player-safe source record, establishing the canonical Learning ownership boundary required by project #75.

---

## Completion note — 2026-09-22

Completed via:

- lisbun/lisjong-arena#348 — source-record v2 allocation provenance, merged as `162df11af1dd15afdf8bce3e72ddfb794e2c096c`
- lisbun/lisjong#186 — Learning L0 vertical slice, merged as `7e5bc0cb1f70268907f0bb30e1a10b1b22e06635`

The canonical path now preserves Arena-owned allocation provenance through:

`PlayerSafeSourceRecord.provenance()["allocation_bindings"]`
→ dataset `manifest["source"]["allocation_bindings"]`
→ model artifact `manifest["source"]["allocation_bindings"]`

Arena remains the allocation authority; lisjong does not query or re-own the live ledger.

**Comment by lisbun** [comment]
Completed.

- Arena dependency: lisbun/lisjong-arena#348 merged (162df11af1dd15afdf8bce3e72ddfb794e2c096c)
- lisjong implementation: #186 merged (7e5bc0cb1f70268907f0bb30e1a10b1b22e06635)
- PR CI was green on both repositories before merge.
- Acceptance Criteria are now checked complete in the Issue body.

The final v2 path preserves Arena-owned allocation provenance into the lisjong dataset/model artifact without introducing Arena-ledger ownership or runtime dependency into lisjong.
